docs/01-tooling/01-automatic-file-bundling.md

```diff
@@ -1,6 +1,188 @@
 ## Automatic JavaScript file bundling and library consumption
+
 ### Intro
+
+With MediaWiki [ResourceLoader][] JavaScript files are run without a module
+system, so files have to expose and consume other files by setting properties
+and objects on global variables, and getting properties and variables from
+global variables.
+
+This has caused issues in the development workflows related to the project's
+JavaScript sources, and the use of JavaScript libraries from the OSS ecosystem.
+
+#### Application JavaScript sources
+
+The order in which this files needs to be loaded so that all dependencies are
+set properly then needs to be [manually specified in
+`extension.json`][scripts-order] by file name.  As such, there are two sources
+of truth:
+
+* The configuration that specifies some order of files,
+* And the source code that implicitly uses files in some order.
+
+If you are doing anything mildly complex, you will end up with a big list of
+JavaScript files that depend on each other. Having to manually understand that
+dependency graph based on the list of dependencies on a JSON file and how the
+files use, define, or re-define global variables from or for other files, can be
+quite of a headache.
+
+Changes to the source code, moving lines in the same file, refactoring code to
+other and new files, adding new code that uses other files, removing code, ...
+
+All of these become really hard really quickly. In the end, organizing your
+code, trying to keep it maintainable, and understanding how it is structured
+have a very big cost on the project's quality, development and maintenance.
+
```

Comment:
I'm going to bold it for now. I feel like it is a crucial point to make, but without the previous explanation I'm not sure it would be taken seriously.

I've added a slightly smaller version to the second paragraph in the intro too.